src/puya/ir/main.py

```diff
@@ -420,7 +420,13 @@ def fold_state_and_special_methods(
     if not (c.init and c.init.body.body):
         result.init = None
     arc4_method_refs = dict(filter(None, maybe_arc4_method_refs.values()))
-    if arc4_method_refs:
+    if arc4_method_refs and not contract.is_arc4:
+        raise CodeError(
```

Comment:
This should be checked earlier during AWST build, and an error is emitted then if and arc4 decorator is used with a non arc4 class, so this might make more sense as an InternalError. To be sure should also add a test for this case to tests/test_expected_output/arc4.test